### ISSUE TITLE
Immediately invoked setState from useEffect/useLayoutEffect causes Maximum call stack size exceeded

### ISSUE THREAD

**Issue by joshgillies** [WebReflection/augmentor#4]
Apologies if this issue submission is lacking details, but I came across this while experimenting with various hooks from React land last night.

From what I can tell it seems like hooks implemented in the following style cause `augmentor` to enter into some race condition.

```js
import augmentor, {useLayoutEffect, useState} from 'https://unpkg.com/augmentor?module'

const a = augmentor(test)
a()

function test() {
  const elapsed = useRaf(5000, 1000) // 🔥 will lock your browser!
  console.log(elapsed)
}

function useRaf (ms, delay) {
  const [elapsed, set] = useState(0)

  useLayoutEffect(() => {
    let raf, timerStop, start

    const onFrame = () => {
      const time = Math.min(1, (Date.now() - start) / ms)
      set(time)
      loop()
    }
    const loop = () => {
      raf = requestAnimationFrame(onFrame)
    }
    const onStart = () => {
      timerStop = setTimeout(() => {
        cancelAnimationFrame(raf)
        set(1)
      }, ms)
      start = Date.now()
      loop()
    }
    const timerDelay = setTimeout(onStart, delay)

    return () => {
      clearTimeout(timerStop)
      clearTimeout(timerDelay)
      cancelAnimationFrame(raf)
    };
  }, [ms, delay])

  return elapsed
}
```

With the limited knowledge I have in this space, it seems as though there are two possible issues here:

 1) `useEffect`/`useLayoutEffect` seem to be triggered regardless of whether the second argument `ref` is the same
 2) calling the update function from `useState` immediately causes `augmentor` to hit some race condition

For reference, here's how the above behaves in React https://codesandbox.io/s/r5rmkr933m

**Comment by joshgillies** [comment]
Just a note, I now recall how I originally came across this issue, as noted in https://github.com/WebReflection/augmentor/issues/3#issuecomment-454651262 when adopting a React hook for `neverland`/`dom-augmentor`/`augmentor` I hit a similar issue to the above, which seems pretty consistent with the assumption that calling an update method from `useState` inside an `useEffect` callback causes issues.

Basically, the example hook implementation for `xstate` https://xstate.js.org/docs/recipes/react.html#hooks had to be changed as follows, otherwise, the race condition was hit.

```diff
function useMachine(machine) {
  // Keep track of the current machine state
  const [current, setCurrent] = useState(machine.initialState);

  // Start the service (only once!)
  const service = useMemo(
    () =>
      interpret(machine)
        .onTransition(state => {
          // Update the current machine state when a transition occurs
          setCurrent(state);
        })
-       .start(),
+       ,
    []
  );

  // Stop the service when the component unmounts
  useEffect(() => {
+   service.start()
    return () => service.stop();
  }, []);

  return [current, service.send];
}
```

Hope this helps!

**Comment by WebReflection** [comment]
there was a very silly bug indeed, I hope it now works as expected. Thanks for filing this.

**Comment by dakom** [comment]
I think you need to re-open... just tried with `neverland 0.5.5` and I still get the bug

**Comment by WebReflection** [comment]
this has been now fixed, please let me know if that's not the case.

**Comment by dakom** [comment]
yup, seems to be fixed (for the useMemo test I posted)! 

Thanks!!!

**Comment by WebReflection** [comment]
> for the useMemo test I posted

to provide some context, it was a huge mistake to invoke right away, so that anything that uses `update()` internally now won't ever be able to create a race and, on top of that, updating many states/refs/memoes at once won't degrade performance anymore, it'll be all changes delivered just in one extra call.

So now these hooks start making sense for real 🎉

TL;DR there won't be races anymore and if any of you find something misaligned with React hooks, please do let me know

**Comment by joshgillies** [comment]
@WebReflection great work, thanks for the quick fix! 🍻
